docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+
+    QUESTION: do we want to codify that group ``path`` end, can end, or must end
+    in ``/`` ?
+
+.. _dimension:
+.. _dimensions:
+
+*Dimension*
+
+    An array_ has a fixed number of zero or more dimensions. Each
+    dimension has an integer length. This specification only considers
+    the case where the lengths of all dimensions are finite. However,
+    `extensions`_ may be defined which allow a dimension to have
+    an infinite or variable length.
+
+.. _shape:
+
+*Shape*
+
+    The shape of an array_ is the tuple of dimension_ lengths. For
+    example, if an array_ has 2 dimensions_, where the length of the
+    first dimension_ is 100 and the length of the second dimension_ is
+    20, then the shape of the array_ is (100, 20). A shape can be the empty
+    tuple in the case of zero-dimension arrays (scalar)
+
+.. _element:
+.. _elements:
+
+*Element*
+
+    An array_ contains zero or more elements. Each element can be
+    identified by a tuple of integer coordinates, one for each
+    dimension_ of the array_. If all dimensions_ of an array_ have
+    finite length, then the number of elements in the array_ is given
+    by the product of the dimension_ lengths. An array_ may not have
+    been fully initialized.
+
+.. _data type:
+
+*Data type*
+
+    A data type defines the set of possible values that an array_ may
+    contain, and a binary representation (i.e., sequence of bytes) for
+    each possible value. For example, the little-endian 32-bit signed
+    integer data type defines binary representations for all integers
+    in the range −2,147,483,648 to 2,147,483,647. This specification
+    only defines a limited set of data types, but extensions
+    may define other data types.
+
+.. _chunk:
+.. _chunks:
+
+*Chunk*
+
+    An array_ is divided into a set of chunks, where each chunk is a
+    hyperrectangle defined by a tuple of intervals, one for each
+    dimension_ of the array_. The chunk shape is the tuple of interval
+    lengths, and the chunk size (i.e., number of elements_ contained
+    within the chunk) is the product of its interval lengths.
+
+    The chunk shape elements are non-zero when the corresponding dimensions of
+    the arrays are of non-zero length.
+
+.. _grid:
+.. _grids:
+
+*Grid*
+
+    The chunks_ of an array_ are organised into a grid. This
+    specification only considers the case where all chunks_ have the
+    same chunk shape and the chunks form a regular grid. However,
+    extensions may define other grid types such as
+    rectilinear grids.
+
+.. _memory layout:
+.. _memory layouts:
+
+*Memory layout*
+
+    An array_ is associated with a memory layout which defines how to
+    construct a binary representation of a single chunk_ by organising
+    the binary values of the elements_ within the chunk_ into a single
+    contiguous sequence of bytes. This specification defines two types
+    of memory layout based on "C" (row-major) and "F" (column-major)
```

Comment:
Rather than specifically support only these two orders, I think it would be better to support any arbitrary permutation, as that provides additional flexibility with no real added implementation complexity.

Additionally, in most cases users would want the chunk key encoding to match the "memory layout", so there should also be a similar order option for the chunk key encoding.

Reply:
The proposal of @alimanfoo and me is to leave both C and F major layouts in, as they are now, mostly to retain backwards-compatibility. I'll add more context in #126 directly. Arbitrary dimension order can be added via an extension, since this is already an explicit extension point, see #129.

See point 16 in https://hackmd.io/@alimanfoo/SJ1DNVRxo

I'll mark this point as resolved for now, but keep #126 open for the moment to track this discussion. Please feel free to comment both in #126 and #129 directly.